include/aws/common/array_list_with_map.h

```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+#define AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/common.h>
+#include <aws/common/hash_table.h>
+
+struct aws_array_list_with_map {
+    struct aws_array_list list;
+    struct aws_hash_table map; /* map from the element to the index in the array */
+};
+
+AWS_EXTERN_C_BEGIN
+
+/* TODO: CBMC verification. */
+/**
+ * Initialize the array_list with map structure, which support constant time of insert, remove and get random element
+ * from the data structure.
+ *
+ * The underlying hash map will use hash_fn to compute the hash of each element. equals_fn to compute equality of two
+ * keys.
+ *
+ * @param list_with_map             Pointer of structure to initialize with
+ * @param allocator                 Allocator
+ * @param hash_fn                   Compute the hash of each element
+ * @param equals_fn                 Compute equality of two keys
```

Comment:
Nit picky, but I like how hash_table.h describes these functions better:
`equals_fn to compute equality of two keys` of each sounds like your iterating through the table.
equals_fun computes the equality of two elements not keys. the fact the the elements are keys in hashtable is an implementation detail.